// entities/fileFacets.js
export const fileFacets = [
  {
    name: 'file_id',
    title: 'File',
    collapsed: false,
    facetType: 'free-text',
    placeholder: 'File name or ID',
    removable: false,
  },
  {
    name: 'data_category',
    title: 'Data Category',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'data_type',
    title: 'Data Type',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'experimental_strategy',
    title: 'Experimental Strategy',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'analysis.workflow_type',
    title: 'Workflow Type',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'data_format',
    title: 'Data Format',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'platform',
    title: 'Platform',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
  {
    name: 'access',
    title: 'Access Level',
    collapsed: false,
    facetType: 'terms',
    removable: false,
  },
];

export default fileFacets;
