// components/AddToCartButtonSingle.js
// Vendor
import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';
import Color from 'color';

// Custom
import { toggleInCart } from 'dux/cart';
import Button from 'uikit/Button';
import theme from 'theme';

/*----------------------------------------------------------------------------*/

const styles = {
  button: {
    padding: '3px 5px',
    margin: '0 auto',
    border: `1px solid ${theme.greyScale4}`,
  },
  inactive: {
    ':hover': {
      backgroundColor: theme.greyScale6,
    },
  },
  active: {
    backgroundColor: theme.success,
    color: 'white',
    ':hover': {
      backgroundColor: Color(theme.success).darken(0.3).rgbString(),
    },
  },
};

const fileInCart = (files, file) => files.some(f => f.file_id === file.file_id);

const AddToCartButtonSingle = ({ dispatch, file, files, addText, removeText, style, ...props }) => {
  const inCart = fileInCart(files, file);
  return (
  <Button
    style={{ ...style, ...styles.button, ...(inCart ? styles.active : styles.inactive) }}
    onClick={() => dispatch(toggleInCart(file))}
    {...props}
  >
    <ShoppingCartIcon style={{ marginRight: addText ? '5px' : 0 }} />
    {inCart ? removeText : addText}
  </Button>
  );
}

AddToCartButtonSingle.propTypes = {
  files: PropTypes.array,
  file: PropTypes.object,
  dispatch: PropTypes.func,
  style: PropTypes.object,
  addText: PropTypes.string,
  removeText: PropTypes.string,
};

/*----------------------------------------------------------------------------*/

export default connect(state => state.cart)(AddToCartButtonSingle);
