// components/EntityPageHorizontalTable.js
import React, { PropTypes } from 'react';

import { Row, Column } from 'uikit/Flex';
import theme from 'theme';
import Table, { Tr, Td, Th } from 'uikit/Table';

// th are horizontal
const EntityPageHorizontalTable = ({ style, title, rightComponent, headings, data, emptyMessage}) => {
  const styles = {
    table: {
      borderCollapse: 'collapse',
      borderSpacing: 0,
      overflow: 'auto',
      backgroundColor: '#fff',
    },
    tr: {
      borderTop: `1px solid ${theme.greyScale5}`,
      borderBottom: `1px solid ${theme.greyScale5}`,
      borderRight: 0,
      borderLeft: 0,
    },
    td: {
      borderRight: 0,
    },
  };
  return (
    <Column
      style={{
        flexWrap: 'wrap',
        marginRight: '2rem',
        overflow: 'scroll',
        borderRight: `1px solid ${theme.greyScale5}`,
        borderTop: `1px solid ${theme.greyScale5}`,
        borderLeft: `1px solid ${theme.greyScale5}`,
        borderBottom: 0,
        ...style,
      }}
    >
      <h3
        style={{
          color: theme.greyScale7,
          width: '100%',
          fontSize: '24px',
          lineHeight: '1.4em',
          fontWeight: 'normal',
          marginTop: 0,
          marginBottom: 0,
          padding: '1rem',
          backgroundColor: '#fff',
          display: 'flex',
          justifyContent: 'space-between',
        }}
      >{title} {rightComponent}</h3>
      {data.length ? (
        <Table
          style={styles.table}
          headings={headings.map(h => <Th key={h.title}>{h.title}</Th>)}
          body={
            <tbody>
            {data.map((d, i) => {
              const headingKeys = headings.map(h => h.key);
              return (
                <Tr
                  style={{
                    ...styles.tr,
                    backgroundColor: i % 2 === 0 ? theme.greyScale6 : '#fff',
                  }}
                  key={i}
                >
                {headingKeys.map(hk => (
                  <Td key={d[hk]}>{d[hk]}</Td>
                ))}
                </Tr>);
            })}
            </tbody>
        }
        />) : (
        <Row
          style={{
            borderBottom: `1px solid ${theme.greyScale5}`,
          }}
        >
          <h4 style={{ padding: '1rem' }}>{emptyMessage}</h4>
        </Row>)
      }
    </Column>
  );
};

EntityPageHorizontalTable.propTypes = {
  title: PropTypes.node,
  style: PropTypes.object,
  rightComponent: PropTypes.object,
  headings: PropTypes.array,
  data: PropTypes.array,
  emptyMessage: PropTypes.string,
};

export default EntityPageHorizontalTable;
