// components/RemoveFromCartButton.js
// Vendor
import React, { PropTypes } from 'react';
import XIcon from 'react-icons/lib/fa/close';
import DownCaretIcon from 'react-icons/lib/fa/caret-down';
import TrashIcon from 'react-icons/lib/fa/trash-o';
import { connect } from 'react-redux';

// Custom
import withDropdown from 'uikit/withDropdown';
import Button from 'uikit/Button';
import { Column, Row } from 'uikit/Flex';
import { dropdown } from 'theme/mixins';
import theme from 'theme';
import { toggleInCart } from 'dux/cart';
import { userCanDownloadFile } from 'utils/auth';

/*----------------------------------------------------------------------------*/

const styles = {
  row: {
    color: theme.greyScale2,
    padding: '0.6rem 1rem',
    ':hover': {
      backgroundColor: theme.greyScale6,
    },
  },
  iconSpacing: {
    marginRight: '0.6rem',
  },
};

const RemoveFromCartButton = ({
  active,
  toggleActive,
  mouseDownHandler,
  mouseUpHandler,
  style,
  files,
  user,
  dispatch,
}) => {
  const unauthorizedFiles = files.filter((file) => !userCanDownloadFile({ user, file }));

  return (
    <Button
      style={style}
      leftIcon={<TrashIcon />}
      rightIcon={<DownCaretIcon />}
      onClick={toggleActive}
      onMouseDown={mouseDownHandler}
      onMouseUp={mouseUpHandler}
    >
      Remove From Cart

      {active &&
        <Column
          style={{ ...dropdown, width: '22rem' }}
        >
          <Row
            style={styles.row}
            onClick={() => dispatch(toggleInCart(files))}
          >
            <XIcon style={styles.iconSpacing} /> All Files({files.length})
          </Row>
          <Row
            style={styles.row}
            onClick={() => dispatch(toggleInCart(unauthorizedFiles))}
          >
            <XIcon style={styles.iconSpacing} /> Unauthorized Files({unauthorizedFiles.length})
          </Row>
        </Column>
      }
    </Button>
  );
};

RemoveFromCartButton.propTypes = {
  active: PropTypes.bool,
  toggleActive: PropTypes.func,
  mouseDownHandler: PropTypes.func,
  mouseUpHandler: PropTypes.func,
  style: PropTypes.object,
  files: PropTypes.array,
  user: PropTypes.object,
  dispatch: PropTypes.func,
};

/*----------------------------------------------------------------------------*/

export default connect(
  (state) => ({ ...state.cart, ...state.auth })
)(withDropdown(RemoveFromCartButton));
