// containers/FilesPage.js
// Vendor
import React, { PropTypes } from 'react';
import Relay from 'react-relay';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';

// Custom
import DownloadManifestButton from 'containers/DownloadManifestButton';
import CasesAggregations from 'containers/CasesAggregations';
import FilesAggregations from 'containers/FilesAggregations';
import FileTable from 'containers/FileTable';
import FileFacets from 'components/FileFacets';
import SearchPage from 'components/SearchPage';
import Button from 'uikit/Button';
import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';
import { addAllPagesToCart } from 'dux/cart';

/*----------------------------------------------------------------------------*/

const FilesPage = ({ viewer, relay, location, dispatch }) => {
  const hits = viewer.files.hits;

  const setAutocomplete = quicksearch => relay.setVariables({
    quicksearch,
    runQuicksearch: !!quicksearch,
  });

  const Aggregations = {
    Cases:
      <CasesAggregations
        aggregations={viewer.cases.aggregations}
        hits={(viewer.cases || {}).hits || {}}
        setAutocomplete={setAutocomplete}
      />,
    Files:
      <FilesAggregations
        aggregations={viewer.files.aggregations}
        hits={(viewer.files || {}).hits || {}}
        setAutocomplete={setAutocomplete}
      />,
  };

  const Results = <FileTable hits={viewer.files.hits} />;
  const Facets = <FileFacets Aggregations={Aggregations} />;
  const Buttons = [
    <Button
      leftIcon={<ShoppingCartIcon />}
      onClick={() => {
        dispatch(addAllPagesToCart({
          total: hits.pagination.total,
          filters: location.query.filters,
        }))
      }}
    >
      Add all files to the cart
    </Button>,
    <DownloadManifestButton hits={hits} />,
  ];

  return (
    <SearchPage
      Facets={Facets}
      Results={Results}
      Buttons={Buttons}
    />
  );
};

FilesPage.propTypes = {
  viewer: PropTypes.object,
  relay: PropTypes.object,
  addPagesToCart: PropTypes.func,
  location: PropTypes.object,
  dispatch: PropTypes.func,
};

export { FilesPage };

export default Relay.createContainer(connect()(withRouter(FilesPage)), {
  initialVariables: {
    first: 20,
    offset: 0,
    filters: null,
    quicksearch: '',
    sort: '',
    runQuicksearch: false,
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Root {
        cases {
          aggregations(filters: $filters) {
            ${CasesAggregations.getFragment('aggregations')}
          }
          hits(first: 5, quicksearch: $quicksearch) @include(if: $runQuicksearch) {
            ${CasesAggregations.getFragment('hits')}
          }
        }
        files {
          aggregations(filters: $filters) {
            ${FilesAggregations.getFragment('aggregations')}
          }
          hits(first: $first, offset: $offset, filters: $filters, sort: $sort) {
            pagination {
              total
            }
            ${DownloadManifestButton.getFragment('hits')}
            ${FileTable.getFragment('hits')}
          }
        }
      }
    `,
  },
});
